### DISCUSSION TITLE
Data Importer "Unable to connect" on docker installation

### DISCUSSION THREAD

**Discussion by yarinbar** [firefly-iii/firefly-iii#8218] [Firefly III]
I followed the installation guide closely:
https://docs.firefly-iii.org/data-importer/installation/docker/
https://docs.firefly-iii.org/data-importer/installation/configuration/#client-id-firefly-iii-url

Created PAT, created client ID put them both into the .importer.env file:
```
FIREFLY_III_URL=http://app:8080
VANITY_URL=http://localhost
FIREFLY_III_ACCESS_TOKEN=eyJ0eXAiOiJKV1QiLCJhbGciO....
FIREFLY_III_CLIENT_ID=16
```

I looked into the terminal logs to determine what was the IP to use on the Client ID url:

```
firefly_iii_core  | 172.19.0.1 - - [01/Dec/2023:08:14:52 +0100] "GET /chart/category/frontpage HTTP/1.1" 200 2890 "-" "Mozilla/5.0 (X11; Linux x86_64; rv:120.0) Gecko/20100101 Firefox/120.0"
```

I figured that i should use 172.19.0.2 (but that sometimes changes when i down docker and then up again - for example it used to be `172.18.0.1` - should i recreate the client each time it changes?)

Also changed the passwords in both files to something different than the default and to the exact same thing.

Firefly works normally, I created a few accounts with IBAN.

When i try to browse to localhost:81 i get Unable to connect  page:

![image](https://github.com/firefly-iii/firefly-iii/assets/37772157/c6ff82ec-1ea6-4248-bf77-5abc100e64ae)

**Comment by JC5** [comment]
So is the port forwarded to 81?

**Reply by yarinbar** [reply]
No sorry, can you help me with that? What command should i run?

And do i have to change the client when the ip changes? 

For example when i see the ip is 172.19.0.1 then use client id with ip 172.19.0.2 and when that changes to 172.16.0.1 change it to 172.16.0.2?

**Reply by Bond246** [reply]
in the docker-compose file from firefly-iii, which you said you used, port 81 is defined and mapped for the importer container.

But you can check this with
`docker container list -f "name=firefly*"` This should list alle firefly-containers. In Ports column there should be somehing like `0.0.0.0:81->8080/tcp`

Normaly it shouldn't be needed to change ip-addresses everytime thats why you define a container-name to reference this as a DNS-name for the container. Both applications, firefly an the importer, should communicate via the internal bridge-network and the DNS-names of the docker-compose-stack and/or the container-name.

**Reply by yarinbar** [reply]
There is no entry with port 81:

```
docker container list -f "name=firefly*"                                                                                                          
CONTAINER ID   IMAGE                    COMMAND                   CREATED       STATUS                 PORTS                                   NAMES
52ebae0142f0   fireflyiii/core:latest   "/usr/local/bin/entr…"    7 hours ago   Up 7 hours (healthy)   0.0.0.0:80->8080/tcp, :::80->8080/tcp   firefly_iii_core
d1f2fc0e2ae5   mariadb                  "docker-entrypoint.s…"    7 hours ago   Up 7 hours             3306/tcp                                firefly_iii_db
a20ada20053b   alpine                   "sh -c 'echo \"0 3 * …"   7 hours ago   Up 7 hours                                                     firefly_iii_cron
```

**Reply by Bond246** [reply]
I'm not shure but it doesn't looks like your importer container is running.
With `docker ps --filter "status=exited"` you see stopped containers.

But maybe it helps more to install [portainer](https://docs.portainer.io/start/install-ce/server/docker) for docker management. There you have a better overview about running (and stopped) containers.

**Comment by SawkeeReemo** [comment]
I'm having a similar problem, but all the links to the documentation for the importer seem to be broken. So I don't even know what I'm supposed to troubleshoot. I can get launch the Importer screen, but as soon as I put in my Client ID it goes to some random address and fails. The address starts like: http://localhost/oauth/authorize?client_id=...

But that's not going to work since I'm running this inside a docker container. (I don't think anyway)... 

And since the docs are all offline, I see this line in the .importer.env file:
`# 3) In case of Docker, refer to the internal IP of your Firefly III installation.`

Does that actually mean the container's internal IP address? Or does it mean the host machine IP? 
![Screen Shot 2023-12-12 at 7 13 40 PM](https://github.com/firefly-iii/firefly-iii/assets/60272670/1be8aac3-df76-42fb-be04-ecbe98aa6337)

**Reply by SawkeeReemo** [reply]
I just kept guessing at things until something worked, and it looks like it's actually the host IP and whatever port you assigned to the `firefly_iii_core` container. Not the internal container IP as the comment in the .env file suggests. (someone please correct me if I'm wrong; I'm literally guessing over here)

`FIREFLY_III_URL=http://my.docker.host.ip:port`

I also set `VANITY_URL` to the same thing since it's not clear what you're supposed to put there. And that seemed to work. I think it having the default `http://localhost` value in there was causing the auth link to fail.

While I was in there, I manually set the Client ID that I set in my profile in Firefly3, then I recreated the `firefly_iii_importer` container.

Hopefully this works (and helps anyone else who's searching)... ¯\\\_(ツ)_/¯

**Reply by JC5** [reply]
Nice, glad to hear you fixed it. Most dead links are on me, the documentation has been rewritten recently. But at the top of the 404 page you'll find a menu, and a search bar. You should definitely try them.

**Reply by SawkeeReemo** [reply]
Will do, thank you! Really excited to get this up and running now that Mint.com has dropped the bomb on us. Just hoping I can get the automated transactions figured out (for U.S.) because I already know I'm going to fail at doing that manually every month. 😂

**Reply by SawkeeReemo** [reply]
Confirmed that the menu links on the site are working... allow me to facepalm over here. And this is exactly what I was trying to solve on my own earlier. (leaving this here in case anyone searching is looking for help)
![Screen Shot 2023-12-12 at 9 09 10 PM](https://github.com/firefly-iii/firefly-iii/assets/60272670/5c13aa13-f889-44a7-8db1-9e495e1d96fc)

**Reply by Bond246** [reply]
In my case i use my own docker-network where i put all my services inside. In this case naming the firefly-app only `app` would intereference with other services that often do the same. 
So what i mostly do is naming it like `firefly-app` and in addition using a custom container-name like `firefly-app`. In this case the `FIREFLY_III_URL` would be `http://firefly-app:8080`.

In general to get more control about that options i would always recommend to use unic container-names and use that container-names in ENVs to connect the services with eath other.
